ocp_resources/utils.py

```diff
@@ -26,27 +27,90 @@ class TimeoutSampler:
+    {
+        exception0: [exception0_msg0],
+        exception1: [
+            exception1_msg0,
+            exception1_msg1
+        ],
+        exception2: []
+    }
+
+    Args:
+        wait_timeout (int): Time in seconds to wait for func to return a value equating to True
+        sleep (int): Time in seconds between calls to func
+        func (function): to be wrapped by TimeoutSampler
+        exceptions (tuple): Deprecated: Tuple containing all retry exceptions to pass to TimeoutSampler
+        exceptions_msg (str): Deprecated: String to match exception against
+        exceptions_dict (dict): Definition for exception handling
+        print_log (bool): Print elapsed time to log
     """
 
     def __init__(
         self,
         wait_timeout,
         sleep,
         func,
-        exceptions=None,
-        exceptions_msg=None,
+        exceptions=None,  # Deprecated
+        exceptions_msg=None,  # Deprecated
+        exceptions_dict=None,
         print_log=True,
         **func_kwargs,
     ):
         self.wait_timeout = wait_timeout
         self.sleep = sleep
         self.func = func
         self.func_kwargs = func_kwargs
-        self.exception = exceptions or Exception
         self.elapsed_time = None
-        self.exceptions_msg = exceptions_msg
         self.print_log = print_log
 
+        self.exceptions_dict = {}
+        self.exception = None  # TODO: Set when exceptions arg removed
+        self.exceptions_msg = None  # TODO: Remove when exceptions_msg arg removed
+        self._pre_process_exceptions(
+            exceptions=exceptions,
+            exceptions_msg=exceptions_msg,
+            exceptions_dict=exceptions_dict,
+        )
+
+    def _pre_process_exceptions(self, exceptions, exceptions_msg, exceptions_dict):
+        """
+        Process exception input for use within _process_execution()
+
+        TODO: Deprecation: This method should be removed when the 'exceptions' argument is removed from __init__
+
+        Validate exception inputs:
+            exceptions
+            exceptions_msg
+            exceptions_dict
+        """
+
+        if exceptions_dict and (exceptions or exceptions_msg):
+            raise ValueError(
+                "Must specify either exceptions_dict or exceptions/exception_msg, not both"
+            )
+
+        elif exceptions or exceptions_msg:
+            warn(
+                "TimeoutSampler() exception and exception_msg are now deprecated. "
+                "Please update to use exceptions_dict by 2021-MM-DD",
+                DeprecationWarning,
+            )
+
+        if exceptions_dict:
+            self.exceptions_dict.update(exceptions_dict)
+            self.exception = tuple(self.exceptions_dict.keys())
+
+        elif exceptions is None:
+            self.exceptions_dict[Exception] = []
+            self.exception = (Exception,)
+
+        else:
+            self.exception = exceptions
+            self.exceptions_msg = exceptions_msg
```

Comment:
You need to process all the args (exception, exception_msg) and create only a dict and work with it
in this function you should return only exceptions_dict

Reply:
@sarahbx as discussed, this will require a refactor. Let's refactor the existing code to use only the dict (and remove self.exception usage)

Reply:
@myakove @rnetser as mentioned previously, we are retaining `self.exception` so we only process the tuple of extensions once. I was not intending to reverse process the input arguments exceptions/exception_msg into exceptions_dict as this code will be removed when exceptions/exceptions_msg arguments are removed.

Reply:
Sure but it is not needed as a separate variable; it is only used once and thus:

`except tuple(self.exceptions_dict.keys()) as exp:`

Reply:
`except tuple(self.exceptions_dict.keys()) as exp:` is nested within a `while`, nested within an `iter`. It will be run more than once.

Reply:
Code updated. Removed `self.exceptions_msg` and only use `self.exceptions_dict` and `self._exceptions` (to ensure we only calculate it once per object)